### ISSUE TITLE
Flux performance spikes on LC machines

### ISSUE THREAD

**Issue by jameshcorbett** [flux-framework/flux-core#2662]
I've been running some tests on rztopaz to get some performance baselines for the UQP, and while doing so I found that Flux suffers from some horrible performance spikes. Sometimes Flux performs on par with Slurm, sometimes it is more than ten times slower. I haven't been able to reproduce the issue on Quartz (there Flux seems to consistently have a slight edge on Slurm).

Here's the terminal session I was using to run my little testing script:
```
[corbett8@rztopaz188:corbett8]$ echo $new_flux 
/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/bin/flux

[the slurm version]
bash-4.2$ python3 uqp/core/ensemble/test/speed_comparison.py 
cmd: ['srun', '--exclusive', '--mpibind=off', '-n1', 'sleep', '1'], resource manager: 'toss', total runs: 200, total cores: 36

managing resources:
Imitating management of 36 total cores
total time: 15.039023399353027 seconds
expected time: 6 seconds
runs per second: 13.29873587460499
Ideal runs per second: 36

loadup:
total time: 14.963631868362427 seconds
expected time: 6 seconds
runs per second: 13.365739130675859
Ideal runs per second: 36

loadup_parallel:
total time: 15.21632432937622 seconds
expected time: 6 seconds
runs per second: 13.143778725449842
Ideal runs per second: 36

timing for one run at a time:
total time for 20 runs: 22.66350483894348
average time: 1.1331752419471741

[the flux version]
bash-4.2$ srun --pty -n1 --mpibind=off --exclusive $new_flux start python3 uqp/core/ensemble/test/speed_comparison.py 
cmd: ['flux', 'mini', 'run', '-n1', '-c1', 'sleep', '1'], resource manager: 'flux', total runs: 200, total cores: 36

managing resources:
Imitating management of 36 total cores
total time: 244.74679040908813 seconds
expected time: 6 seconds
runs per second: 0.8171710839014682
Ideal runs per second: 36

loadup:
total time: 239.944030046463 seconds
expected time: 6 seconds
runs per second: 0.8335277187820501
Ideal runs per second: 36

loadup_parallel:
total time: 239.984849691391 seconds
expected time: 6 seconds
runs per second: 0.833385941892542
Ideal runs per second: 36

timing for one run at a time:
total time for 20 runs: 25.631010055541992
average time: 1.2815505027770997
2020-01-21T17:34:45.415953Z broker.err[0]: rc3: flux-module: cmb.rmmod[0] sched-simple: No such file or directory
```

The contents of speed_comparison.py:
```
"""
This script is meant to provide a performance baseline for the UQP's ensemble manager.

The various timing tests roughly mimic the ensemble manager's behavior.
"""
import subprocess as sp
import multiprocessing as mp
import time
import sys
import os
import math
from concurrent import futures


def loadup_cmds_parallel(cmd, total_runs, total_cores):
    """Open `total_runs` processes executing `cmd` as quickly as possible in parallel, and time their completion."""
    executor = futures.ThreadPoolExecutor(10)
    cmds_iterable = [cmd for _ in range(total_runs)]
    start_time = time.time()
    procs = executor.map(launch_devnull, cmds_iterable)
    future_list = []
    for p in procs:
        future_list.append(executor.submit(p.wait))
    futures.wait(future_list)
    total_time = time.time() - start_time
    assert(all(p.returncode == 0 for p in procs))
    print_info(total_runs, total_time, total_cores)
    executor.shutdown()
    return total_time


def loadup_cmds(cmd, total_runs, total_cores):
    """Open `total_runs` processes executing `cmd` as quickly as possible serially, and time their completion."""
    start_time = time.time()
    procs = [launch_devnull(cmd) for _ in range(total_runs)]
    for p in procs:
        p.wait()
    total_time = time.time() - start_time
    assert(all(p.returncode == 0 for p in procs))
    print_info(total_runs, total_time, total_cores)
    return total_time


def manage_resources(cmd, total_runs, total_cores):
    """Open `total_runs` processes executing `cmd` and time their completion.

    Ensure that there are never more than `total_cores` processes executing `cmd` at a time.
    (cmd is assumed to be a single-core application).
    """
    launcher = get_launcher()
    orig_job_count = total_runs
    print("Imitating management of {} total cores".format(total_cores))
    concurrent_runs = min(total_cores, total_runs)
    start_time = time.time()
    jobs = [launcher(cmd) for i in range(concurrent_runs)]
    run_counter =  total_runs - concurrent_runs
    while run_counter > 0:
        for index, job in enumerate(jobs):
            if run_counter <= 0:
                break
            elif job.poll() is not None:
                run_counter -= 1
                jobs[index] = launcher(cmd)
    for job in jobs:
        job.wait()
    total_time = time.time() - start_time
    print_info(total_runs, total_time, total_cores)
    return total_time


def print_info(total_runs, total_time, total_cores):
    """Print summary info about run time."""
    print("Total time: {} seconds".format(total_time))
    print("Ideal time: {} seconds".format(math.ceil(total_runs/total_cores)))
    print("Runs per second: {}".format(total_runs/total_time))
    print("Ideal runs per second: {}".format(total_cores))


def get_launcher(with_file=False):
    """Return a function that will launch commands in other processes."""
    if with_file:
        jobs_index = 0
        def launch_with_file(args):
            """Open a file and redirect stdout and stderr there."""
            nonlocal jobs_index
            with open("placeholder{}.log".format(jobs_index), "a") as run_log:
                process = sp.Popen(args, stdout=run_log, stderr=sp.STDOUT, text=True)
            jobs_index += 1
            return process
        return launch_with_file
    else:
        return launch_devnull


def launch_devnull(cmd):
    """Execute cmd in a separate process, redirecting stdout and stderr to devnull"""
    return sp.Popen(cmd, stdout=sp.DEVNULL, stderr=sp.DEVNULL)


def one_at_a_time(cmd, total_runs):
    """Time commands in succession, and return the average time"""
    start_time = time.time()
    for i in range(total_runs):
        sp.run(cmd, stdout=sp.DEVNULL, stderr=sp.DEVNULL)
    total_time = time.time() - start_time
    print("total time for {} runs: {}".format(total_runs, total_time))
    print("average time: {}".format(total_time / total_runs))
    return total_time


def main():
    """Run a mini suite of performance tests"""
    # initialize these variables to the right values for your setup
    resource_manager = "toss"
    total_runs = 200
    total_cores = 36
    application = ["sleep", "1"]
    # now let the rest happen...
    if resource_manager == "flux":
        cmd = ["flux", "mini", "run", "-n1", "-c1"]
    elif resource_manager == "blueos":
        cmd = ["lrun", "--pack", "-n1", "--mpibind=off"]
    elif resource_manager == "toss":
        cmd = ["srun", "--exclusive", "--mpibind=off", "-n1"]
    elif resource_manager is None:
        cmd = []
    else:
        raise ValueError()
    cmd.extend(application)
    print("cmd: {}, resource manager: '{}', total runs: {}, total cores: {}".format(cmd, resource_manager, total_runs, total_cores))
    print("\nmanaging resources:")
    manage_resources(cmd, total_runs, total_cores)
    print("\nloadup:")
    loadup_cmds(cmd, total_runs, total_cores)
    print("\nloadup_parallel:")
    loadup_cmds_parallel(cmd, total_runs, total_cores)
    print("\ntiming for one run at a time:")
    one_at_a_time(cmd, min(50, total_runs // 10))


if __name__ == '__main__':
    main()
```

I'm going to try out as many other machines as I can to see if I see anything similar.

**Comment by dongahn** [comment]
@jameshcorbett: this is good. (not that slow down is good but that we have another person testing and reporting this.) I will probably want to reproduce this (after working on `flux-tree`). If this is a scheduler, optimization is part of my future work item any way.

**Comment by jameshcorbett** [comment]
Sounds good, thanks @dongahn. I did confirm that the same performance degradation happens on Quartz, not just rztopaz. I guess I just hadn't run often enough on Quartz to see the issue before.

**Comment by grondo** [comment]
It would be interesting if you still see the variability without the graph scheduler loaded, e.g. using the `simple-sched` fcfs scheduler in flux-core. This would determine if the issue needs to be opened against flux-sched or flux-core.

The simple scheduler may actually be slower than flux-sched, so you'd just want to be looking for unexplained variation.

**Comment by dongahn** [comment]
@grondo: agreed.

**Comment by jameshcorbett** [comment]
I'd be happy to try that, @grondo; how do I specify the scheduler plugin?

**Comment by grondo** [comment]
I think you can set `FLUX_QMANAGER_RC_NOOP=t` in your environment to disable the load of the `qmanager` scheduler, and leave the `sched-simple` module in place.

Verify with `flux module list`:

```console
$ $flux start flux module list | grep sched 
qmanager              1027184 3E7D15F    0  S sched
$ FLUX_QMANAGER_RC_NOOP=t $flux start flux module list | grep sched
sched-simple          1229640 32F8013    0  S sched

```

**Comment by jameshcorbett** [comment]
Sorry for the delay--I stopped focusing on performance for a while. The UQP software lead wants to do a comparison of Slurm and Flux, though, so I've been running some more performance tests, and I've gotten some strange results:

1. It doesn't seem to matter whether I use sched-simple or the graph scheduler; I got similar performance from both.

2. Allocating additional nodes did not seem to give much of a performance improvement. For instance, a 2-node allocation would complete N jobs around 1.16x faster, and a 3-node allocation 1.3x faster, than a single-node allocation. I did, however, manage to get better scaling by splitting my allocation using `flux tree`, so that each Flux instance manages a single node. 

3. On a single node, Flux has a consistent, significant edge on Slurm. After a single node, though, it's Slurm all the way (I haven't used `flux tree` enough to say how much it helps, but my preliminary results say it doesn't make up for the difference). Slurm has, in my tests, scaled linearly up to 12 nodes; I haven't tested beyond that point.

4. In all my Flux benchmarks, submitting all the jobs all at once was consistently ~20% slower than maintaining a sliding window of jobs active (the window size was always equal to the number of allocated cores). The opposite was the case with Slurm, where submitting all the jobs at once always has a slight edge on the sliding window version; a priori I would have expected that to be the case with both Slurm and Flux.

5. I haven't seen the same kind of performance spikes I saw before; I'm not sure if it's because I'm using a new version of Flux or because I just hadn't run enough tests to see a pattern before.

6. The CLI consistently had a slight edge on the bindings.

About the setup: for all the results I described above, I was using `/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.14.x-20190129-ftree/` which @dongahn installed for me. All the tests I ran were just timing the throughput of some number of `sleep 1` applications, each allocated a single core. This test is, of course, ridiculous--I think the fastest jobs the UQP runs are on the order of minutes--and if there's a better benchmark you think I should run, I'd be happy to do it.

Anyway, when I transitioned from my benchmarks to my production code, I noticed a greater discrepancy between Slurm and Flux than the benchmarks led me to expect. The main difference between the benchmarks and my production code is that my production code is busier: in addition to submitting jobs, it's running a bunch of other (GIL-bound, but still) threads which are polling for user input, grabbing new runs from a database, saving state, and running a server to distribute runs. This made me wonder whether Flux isn't "slow"--my script is just not getting the CPU time it needs. So, I thought I would ask--what happens to `myscript.py` when I execute `srun ... flux start myscript.py`? On Slurm I can give `myscript.py` (more or less) dedicated access to a core. Can I do anything similar with Flux? I've been led to believe that I can't, or at least that it isn't easy...

**Comment by dongahn** [comment]
@jameshcorbett: thank you for the report. Maybe we can hash this out a bit at our 4PM meeting today if you can make it?

**Comment by jameshcorbett** [comment]
@dongahn absolutely, I was hoping we could.

**Comment by grondo** [comment]
@jameshcorbett: In my testing a single instance of Flux is limited to running about 50 jobs/s at the moment. It isn't surprising that we'd be a low slower than Slurm at this point since we haven't started optmizing, but if your raw numbers are a lot slower I'd be interested in seeing the results.

>  So, I thought I would ask--what happens to myscript.py when I execute srun ... flux start myscript.py? On Slurm I can give myscript.py (more or less) dedicated access to a core. Can I do anything similar with Flux? I've been led to believe that I can't, or at least that it isn't easy...

Yeah, with `srun flux start myscript.py` your script is running as the "initial program" of the instance and will be sharing a core with the result of the system. As a test why don't you try:
```
srun [OPTIONS] flux start flux mini run -n1 myscript.py
```
This will run `myscript.py` on one core of your instance so it shouldn't be assigned to other jobs. If you want to assign more than one core to the script, add ` -c N` to the `flux mini run` cmdline.

**Comment by jameshcorbett** [comment]
Thanks @grondo! I knew you hadn't started optimizing yet, but I thought I was supposed to expect more than 50 jobs/s. If I'm just reporting what you already know, it doesn't really matter that much.

I also thought I couldn't submit jobs from a script that was the target of a `mini run`. A misunderstanding from a long time ago, I guess! I tried it and it worked very well. That's what I get for not actually trying it myself, I guess.

**Comment by dongahn** [comment]
Thanks @jameshcorbett. 

When we write our paper, it should be apparent about the performance advantages of Flux's hierarchical scheduling. 

Our collaboration gave me a lot of good insight about Flux's scheduling model on large simulation campaign life cycle! Great stuff.
